src/pip/_internal/operations/prepare.py

```diff
@@ -374,6 +383,9 @@ def _fetch_metadata_only(
                 "Metadata-only fetching is not used in the legacy resolver",
             )
             return None
+        # TODO: this should probably not be checked here, but rather later once we
+        # download the wheels? Otherwise hash checking won't be able to use
+        # metadata-only dists at all.
```

Comment:
@pradyunsg:

> One thing about `fast-deps` -- there's low-hanging fruit there seemingly, based on the final few things stated in #8670 that I've not looked into.

After hacking away last night and coming up with the new `.is_concrete` distinction between dists in this change, I'd like to move further optimizations into a separate change. The chunking from #8670 and the `require_hashes` checking here might be good candidates. Does that make sense?

Reply:
Split this out into #12192.